### PR TITLE
[release/6.0.x] Add System.Diagnostics.StackFrame.GetMethodInfoFromNativeIP API for VS4Mac

### PR DESCRIPTION
**Pull request by mikem8361**
# Customer Impact

VS4Mac needs a way to symbolize managed IPs when rethrowing a native NSException from Objective C++. This is a short term API needed for the next 6.0 service release discoverable only through reflection.

Issue: #61186

# Testing

Local testing.  Verified by the VS4Mac team.

# Risk

Low.

### CODE REVIEW

**Review by jkotas — APPROVED** [review]

File: src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml
Comment (MichalStrehovsky) [inline src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml:40]: Why do we need to root this? It doesn't look like the runtime relies on the presence of this method.

**Comment by mikem8361** [comment]
VS4Mac uses reflection to get to and call it and if it isn't in the linker file it gets removed.

mikem
________________________________

**Comment by MichalStrehovsky** [comment]
> VS4Mac uses reflection to get to and call it and if it isn't in the linker file it gets removed.

I assume this happens during the repo build because the API is not public. ILLink.Descriptors.Shared.xml are used both at repo build time and end user app build time. We don't want to root an unused API in _all .NET apps_ just because VS4Mac uses it.

This directive should go in the LibraryBuild.xml file like here:

https://github.com/dotnet/runtime/blob/7cd8f19225f2260017f65886f061fe15973a6efc/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems#L53

LibraryBuild directives are only used when building the repo. They will not root things when building/trimming end user apps.

Cc @eerhardt

**Comment by eerhardt** [comment]
> VS4Mac uses reflection to get to and call it

Can we instead propose a public API for this? That is MUCH better than using private reflection.

**Comment by mikem8361** [comment]
>> VS4Mac uses reflection to get to and call it
>>
> Can we instead propose a public API for this? That is MUCH better than using private reflection.

I thought adding a new public API to a service release wasn't possible and this is temporary until we have time to design a public one. This needs to go into 6.0.1. 

Or are you saying that if I make it public I won't need the ILLink.Descriptors.Shared.xml entry to keep it from being elided?

**Comment by mikem8361** [comment]
@noahfalk can you approve this 6.0 PR also? I'll make the same changes as the 7.0 one you reviewed.

**Comment by eerhardt** [comment]
I didn't notice initially that this is targeting `6.0`. For `7.0` I think we should try to make a full public API instead of reflection.

> Or are you saying that if I make it public I won't need the ILLink.Descriptors.Shared.xml entry to keep it from being elided?

I guess making it `public` in the implementation assembly, and leaving it out of the `ref` assembly would actually work. The linker wouldn't trim it during the repo build. But I think @MichalStrehovsky's suggestion above is better - to add the ILLink.Descriptors entry into this file:

https://github.com/dotnet/runtime/blob/3dfdcb918b515e130f6dc69335886d7685612db9/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.LibraryBuild.xml#L2-L6

And also adding a comment to the entry stating why it is there.

See https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/ILLink-files.md#illinkdescriptorsxml for more information on how to use these ILLink files.

**Comment by jkotas** [comment]
> Or are you saying that if I make it public I won't need the ILLink.Descriptors.Shared.xml entry to keep it from being elided?

Yes, you can make it public in CoreLib, without adding it to the reference assembly. You won't need to mess with the linker directives if you do it that way. Instead, you will need to suppress warning that complains about the implementation and reference assembly being out of sync.

File: src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml
Comment (eerhardt) [inline src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml:39]: (nit) can this unnecessary change be reverted?

**Review by eerhardt — COMMENTED** [review]
Does it make sense / is it possible to add tests?

**Review by jeffschwMSFT — APPROVED** [review]
Approved.  We should take this for consideration in 6.0.x

**Comment by mikem8361** [comment]
There is no easy way to test this API.  I have something that only works on Linux using the native libunwind library to get IPs to  check but it won't work on OSX (libunwind isn't discoverable and the simple unwind function doesn't exist) and Windows.

**Review by eerhardt — APPROVED** [review]
ILLink changes look good to me.

**Comment by Anipik** [comment]
@mikem8361 is this ready to merge or do we need any review here ?

**Comment by jeffschwMSFT** [comment]
@Anipik let me follow-up with @tommcdon.

**Comment by mikem8361** [comment]
Jan: The API that we want to make public should have different shape (it should be StackFrame factory). It requires refactoring that would not appropriate for servicing, and that would not be possible to pull off in the few days that this have to be done in.

**Comment by jeffschwMSFT** [comment]
@mikem8361 have we looked at the failing CI?  With that @Anipik we are ready to merge.

**Comment by mikem8361** [comment]
The failing CI all look like timeouts/existing infrastructure problems.  I'm still waiting for some feedback from Kyle before we merge it.

**Comment by mikem8361** [comment]
We are ready to merge.

**Pull request merged**
